tunneler/tunneler.py

```diff
@@ -50,6 +85,55 @@ def repackage_frame(self, raw_frame):
 
         return parsed_ether_header.get_raw_header() + raw_frame[EtherHeader.TOTAL_HEADER_LEN:]
 
+    def disrupt_layers(self, raw_data):
+        """
+        Take in raw data, parse it protocol layers, and apply this tunnel's service disruption rules to each layer
+        :param raw_data: The raw data of the frame (all data, including layer 2)
+        :return: the raw bytes data of the disrupted frame
+        """
+        recv_etherheader = EtherHeader.parse_header(raw_data[:EtherHeader.TOTAL_HEADER_LEN])
+
+        l2_payload = raw_data[recv_etherheader.length():]
+        recv_etherheader, l2_payload = self._disruption_rules[2](recv_etherheader, l2_payload)
+
+        raw_ip_header = l2_payload[:IpHeader.DEFAULT_HEADER_SIZE]
+        recv_ipheader = IpHeader.parse_header(raw_ip_header)
+        l3_payload = l2_payload[recv_ipheader.length():]
+        recv_ipheader, l3_payload = self._disruption_rules[3](recv_ipheader, l3_payload)
+
+        if recv_ipheader.proto == IPPROTO_UDP:
+            recv_l4_header = UdpHeader.parse_header(l3_payload[:UdpHeader.UDP_HEADER_SIZE])
+            recv_l4_header.pseudo_header = generate_pseudo_header(
+                recv_ipheader.src_ip,
+                recv_ipheader.dst_ip,
+                recv_ipheader.proto,
+                recv_ipheader.tot_len
+            )
+        elif recv_ipheader.proto == IPPROTO_TCP:
+            recv_l4_header = TcpHeader.parse_header(l3_payload[:TcpHeader.DEFAULT_TCP_HEADER_SIZE])
+            recv_l4_header.pseudo_header = generate_pseudo_header(
+                recv_ipheader.src_ip,
+                recv_ipheader.dst_ip,
+                recv_ipheader.proto,
+                recv_ipheader.tot_len
+            )
```

Comment:
take recv_ip_header as parameter to generate_pseudo_header - much better than duplicated unpacking going in now